### PR TITLE
Ellipse intersection

### PR DESCRIPTION
**Pull request by rupakbajgain**
- The ellipse intersection is working when the ellipse is circle. So i think the problem is the eqn of ellipse.

### CODE REVIEW

File: lckernel/cad/math/intersect.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ bool Intersect::operator()(const lc::geo::Vector& v1, const lc::geo::Vector& v2)
 }
 
 bool Intersect::operator()(const lc::geo::Vector& v, const lc::entity::Point& p) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]: All the intersection methods with `Point` has the same body. You should create a method with `geo::Coordinate` and `geo::Base`, so you have to move the `nearestPointOnEntity` method in `geo::Base` (it's only not defined in `Coordinate` and `Area`).

File: lckernel/cad/math/intersect.cpp [hunk 2]
```diff
@@ -128,14 +118,18 @@ bool Intersect::operator()(const lc::geo::Circle& c, const lc::entity::Arc& arc)
...
 
-bool Intersect::operator()(const lc::geo::Circle& c, const lc::entity::Ellipse& e) {
-    // TODO: test if point is on path
-    auto &&coords = maths::Intersection::QuadQuad(c.equation(), e.equation());
-    if (!coords.empty()) {
-        std::cerr << __PRETTY_FUNCTION__ << " TODO Check if point's are on path" << std::endl;
-    }
-    for (auto &i : coords) {
-        _intersectionPoints.push_back(i);
+bool Intersect::operator()(const lc::geo::Circle& c, const lc::entity::Ellipse& arc) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]: Don't call your `Ellipse` `arc` :)
```suggestion
bool Intersect::operator()(const lc::geo::Circle& c, const lc::entity::Ellipse& ellipse) {
```

File: lckernel/cad/math/intersect.cpp [hunk 3]
```diff
@@ -270,9 +250,22 @@ bool Intersect::operator()(const lc::entity::Ellipse& e, const lc::entity::Splin
     return false;
 }
 
-bool Intersect::operator()(const lc::entity::Ellipse& e, const lc::entity::LWPolyline& lwp) {
-    std::cerr << __PRETTY_FUNCTION__ << " requires implementation" << std::endl;
-    return false;
+bool Intersect::operator()(const lc::entity::Ellipse& a, const lc::entity::LWPolyline& lwp) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]:
```suggestion
bool Intersect::operator()(const lc::entity::Ellipse& e, const lc::entity::LWPolyline& lwp) {
```

File: lckernel/cad/math/intersect.cpp [hunk 4]
```diff
@@ -378,6 +371,24 @@ void Intersect::geovisit(const geo::Vector& v, const geo::Arc& arc) {
     }
 }
 
+void Intersect::geovisit(const geo::Vector& v, const geo::Ellipse& arc) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]:
```suggestion
void Intersect::geovisit(const geo::Vector& v, const geo::Ellipse& ellipse) {
```

File: lckernel/cad/math/intersect.cpp [hunk 5]
```diff
@@ -397,6 +408,62 @@ void Intersect::geovisit(const geo::Arc& arc1, const geo::Arc& arc2) {
     }
 }
 
+void Intersect::geovisit(const geo::Arc& arc1, const geo::Ellipse& arc2) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]:
```suggestion
void Intersect::geovisit(const geo::Arc& arc, const geo::Ellipse& ellipse) {
```

File: lckernel/cad/math/intersect.cpp [hunk 6]
```diff
@@ -397,6 +408,62 @@ void Intersect::geovisit(const geo::Arc& arc1, const geo::Arc& arc2) {
...
+            double a2 = (point - arc2.center()).angle();
+            if (Math::isAngleBetween(a1, arc1.startAngle(), arc1.endAngle(), arc1.CCW()) &&
+                arc2.isAngleBetween(a2)) {
+                _intersectionPoints.push_back(point);
+            }
+        }
+    }
+}
+
+void Intersect::geovisit(const geo::Ellipse& arc2, const geo::Arc& arc1) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]:
```suggestion
void Intersect::geovisit(const geo::Ellipse& ellipse, const geo::Arc& arc) {
```

File: lckernel/cad/math/intersect.cpp [hunk 7]
```diff
@@ -397,6 +408,62 @@ void Intersect::geovisit(const geo::Arc& arc1, const geo::Arc& arc2) {
...
+            double a2 = (point - arc2.center()).angle();
+            if (arc2.isAngleBetween(a2) &&
+                Math::isAngleBetween(a1, arc1.startAngle(), arc1.endAngle(), arc1.CCW())) {
+                _intersectionPoints.push_back(point);
+            }
+        }
+    }
+}
+
+void Intersect::geovisit(const geo::Ellipse& arc1, const geo::Ellipse& arc2) {
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]:
```suggestion
void Intersect::geovisit(const geo::Ellipse& ellipse1, const geo::Ellipse& ellipse2) {
```

File: lckernel/cad/math/intersect.cpp [hunk 8]
```diff
@@ -270,9 +250,22 @@ bool Intersect::operator()(const lc::entity::Ellipse& e, const lc::entity::Splin
...
+    // Once added, we can get rid of the dynamic_pointer_casts and simply
+    // call entity1.visit(entity2);
+    for (auto &entity1 : list1) {
+        if (auto arc = std::dynamic_pointer_cast<const lc::geo::Arc>(entity1)) {
+            geovisit(a, *arc.get());
+        }
+        else {
+            geovisit(*std::dynamic_pointer_cast<const lc::geo::Vector>(entity1).get(), a);
+        }
+    }
```
Comment (feragon) [inline lckernel/cad/math/intersect.cpp]: I like this strategy. However, if you want to get rid of the `dynamic_pointer_casts`, you should, for each entity of the polyline, create an intersection and add the result to the intersection points.
In this case, you can change this function to take a `LWPolyline` and a `Visitable`.

**Comment by rupakbajgain** [comment]
The point intersection didn't worked that way,  because Intersection(geo:coordinate,geo:base) And (lc:visitable,lc:visitable) have same priority causing ambiguousness.

**Pull request closed without merging**